src/main/java/executor/service/service/impl/scenario/ScenarioResultServiceImpl.java

```diff
@@ -0,0 +1,194 @@
+        for (ScenarioResult scenarioResult : scenarioResults) {
+            Set<StepResult> collect = scenarioResult.getStepsResults().stream()
+                    .sorted(Comparator.comparing(StepResult::getId))
+                    .collect(Collectors.toCollection(LinkedHashSet::new));
+            scenarioResult.setStepsResults(collect);
+        }
+        return scenarioResults;
+    }
+
+    @Override
+    @Transactional(readOnly = true)
+    public List<ScenarioResultResponse> getAllScenarioResultsByUserId(Integer userId) {
+        List<ScenarioResult> scenarioResults = scenarioResultRepository.findAllByUserId(userId);
+        List<ScenarioResult> sortedScenarioResults = sorted(scenarioResults);
+        return sortedScenarioResults
+                .stream()
+                .map(ScenarioResultResponse::formScenarioResult)
+                .toList();
+    }
+
+    private List<ScenarioResult> sorted(List<ScenarioResult> scenarioResults) {
+        for (ScenarioResult scenarioResult : scenarioResults) {
+            Set<StepResult> collect = scenarioResult.getStepsResults().stream()
+                    .sorted(Comparator.comparing(StepResult::getId))
+                    .collect(Collectors.toCollection(LinkedHashSet::new));
+            scenarioResult.setStepsResults(collect);
+        }
+        return scenarioResults;
+    }
+
+    @Override
+    @Transactional
+    public void deleteById(Integer scenarioId) {
+        Set<StepResult> stepResults = deleteStepResults(scenarioId);
+        deleteSteps(stepResults);
+        scenarioResultRepository.deleteById(scenarioId);
+    }
+
+    private Set<StepResult> deleteStepResults(Integer scenarioId) {
+        Set<StepResult> stepResults = getStepResults(scenarioId);
+        stepResultRepository.deleteAll(stepResults);
+        return stepResults;
+    }
+
+    private Set<StepResult> getStepResults(Integer scenarioId) {
+        return getScenarioResult(scenarioId).getStepsResults();
+    }
+
+    private ScenarioResult getScenarioResult(Integer scenarioId) {
+        return scenarioResultRepository.findById(scenarioId)
+                .orElseThrow(() -> ScenarioResultExceptions.scenarioNotFound(scenarioId));
+    }
+
+    private void deleteSteps(Set<StepResult> stepResults) {
+        stepRepository.deleteAll(getStepsForDelete(stepResults));
+    }
+
+    private List<Step> getStepsForDelete(Set<StepResult> stepResults) {
+        List<Step> stepsForDelete = new ArrayList<>();
+
+        Set<Step> steps = stepResults.stream()
+                .map(StepResult::getStep)
+                .collect(Collectors.toSet());
+
+        List<StepResult> stepResultsByStep;
+        for (Step step : steps) {
+            stepResultsByStep = stepResultRepository.findAllByStep(step);
+            if (stepResultsByStep.size() == 0) {
+                stepsForDelete.add(step);
+            }
+        }
+        return stepsForDelete;
+    }
+}
```

Comment:
Do you think to simplify  this  class?
You can organize the logging using only two tables with one-to-many relationship: scenarios and actions.
In action you can add four or five fields: 
scenario, action, value, result

I think it will give almost all information what necessary to client.

Reply:
I don't agree with you, we need to have a DataBase in third normal form.

Reply:
And why my variant isn't third normal form?